### DISCUSSION TITLE
restic Daemonset hostpath issue

### DISCUSSION THREAD

**Discussion by vijaynetha** [vmware-tanzu/velero#4217] [General discussions]
Hi team,

iam using velero to take backup and restore.
while taking pvc backup it is failing. because restic daemonset is not running.

restic daemonset is not running because of Hostpath issue.
we don't have access to hostpath in my  kubernetes cluster.

1.  why we are giving hostpath in daemonset restic.
2. why we are running restic as daemonset.

Events:
  Type     Reason        Age                 From                  Message
  ----     ------        ----                ----                  -------
  Warning  FailedCreate  11s (x14 over 52s)  daemonset-controller  Error creating: pods "restic-" is forbidden: PodSecurityPolicy: unable to admit pod: [spec.volumes[0]: Invalid value: "hostPath": hostPath volumes are not allowed to be used]

**Comment by sseago** [comment]
To take a backup of a volume mounted by a pod, the Restic pod mounts "/var/lib/kubelet/pods" from the node as a hostpath Volume. This gives the restic pod access to all volumes mounted by pods running on the same node. This is also why Restic must run as a daemonset. We need a restic pod on each node, so when a pod has volumes that needs a restic backup (or restore) the restic pod running on that node handles the task. If we didn't run Restic as a daemonset, we wouldn't guarantee that there was an available pod running that could back up or restore every possible volume mounted by a running pod.

**Comment by vijaynetha** [comment]
thank Q. 

we can't access hostpath in kubernertes cluster.

is there any other way to take backup only PVC without mounting hostpath in restic pod.

**Comment by sseago** [comment]
Velero provides two ways of copying PV data:
1) snapshots -- uses underlying storage snapshots for ebs and other volume types that support snapshots (and have a velero plugin for them)
2) restic -- filesystem copy that works with any storage type

It's generally best to use snapshots when you can, but there are certain restrictions -- the storage provider must support them, there must be a velero storage/snapshotter plugin registered for the type, a volume snapshot location must be configured for the backup, and the region must be the same for backup and restore.

When these requirements can't be met, restic is the fallback option, but as noted above, this requires the restic daemonset running with hostpath volumes. If you can't do this in your environement, you'll have to use the snapshot approach.
